ddtrace/llmobs/_evaluators/runner.py

```diff
@@ -66,12 +64,17 @@ def start(self, *args, **kwargs):
             return
         super(EvaluatorRunner, self).start()
         logger.debug("started %r to %r", self.__class__.__name__)
-        atexit.register(self.on_shutdown)
 
-    def stop(self, *args, **kwargs):
-        if self.status == service.ServiceStatus.STOPPED:
-            return
-        super(EvaluatorRunner, self).stop()
+    def _stop_service(self) -> None:
+        """
+        Ensures all spans are evaluated & evaluation metrics are submitted when evaluator runner
+        is stopped by the LLM Obs instance
+        """
+        self.periodic(_wait_sync=True)
+        self.executor.shutdown(wait=True)
+        # flush remaining evaluation spans & evaluations
+        self.llmobs_service._instance._llmobs_span_writer.periodic()
+        self.llmobs_service._instance._llmobs_eval_metric_writer.periodic()
```

Comment:
hmm having this circular reference back to the service isn't great for maintainability. Having more things depend on the underlying service instance and its internal attributes is going to hurt the ability to change this code down the line so we should try to avoid it if possible.

Could we do this from the top down? It seems like this should be resolved by ordering the shutdowns like you did below in the LLMObs service. Shut down the evaluator and then shutdown the span writer and eval metric writer... so this code here shouldn't be necessary?

Reply:
Good point - i'm going to move it the `periodic` calls back inside the LLMObs `_stop_service` function. I was concerned that these functions would be unnecessarily called if the evaluator runner was never started, but a service status error would be thrown/caught in this case in which the periodic calls would be skipped

>Shut down the evaluator and then shutdown the span writer and eval metric writer... so this code here shouldn't be necessary?
the shutdown function for the writers doesn't actually flush out spans/eval metrics; there's a seperate `on_shutdown` function registered using `atexit` for flushing on process exit, so i think the periodic calls here are needed

Reply:
Going to go ahead and merge this into the main faithfulness implementation branch; agree that the llm obs service should manage order of runner/writer shutdowns + periodic calls. If you see issues with manually calling `periodic` for the writer's, we can address that in the primary faithfulness PR